### PR TITLE
Issue #12087: Create IDEA Inspections Docker image for CircleCI

### PR DESCRIPTION
**Pull request by nrmancuso**
#12087 

Follow up: https://github.com/checkstyle/contribution/issues/669

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -212,28 +212,51 @@ jobs:
       - run:
           command: ./.ci/validation.sh git-diff
 
+  run-inspections:
+    docker:
+      - image: nrmancuso/idea-docker:1.0.0
```
Comment (romani) [inline .circleci/config.yml:217]: Why not a container that is available already in open source?

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: .circleci/config.yml (hunk 1 above)
Comment (nrmancuso) [reply .circleci/config.yml:217]: Couldn't find one that was up to date and/or didn't have bugs. This one is open source, too, with a reliable maintainer :)

Comment (romani) [reply .circleci/config.yml:217]: I am questioning reason of special docker creation. As it is already done by others.

File: config/intellij-idea-inspections.properties [hunk 2]
```diff
@@ -12,4 +12,9 @@ idea.exclude.patterns=.idea/**;\
               src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/ParseTreeBuilder.java;\
               src/test/java/com/puppycrawl/tools/checkstyle/grammar/javadoc/ParseTreeBuilder.java;\
               config/idea.properties;\
-              config/intellij-idea-inspections.properties
+              config/intellij-idea-inspections.properties;\
+              .ci-temp/**;\
+              .classpath;\
+              .project;\
+              .settings/**;\
+              bin/**
```
Comment (nrmancuso) [inline config/intellij-idea-inspections.properties:20]: these were added to make sure that scope in project matches this file, since we cannot specify scope from cli.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java [hunk 3]
```diff
@@ -3633,7 +3633,7 @@ public final class TokenTypes {
     public static final int LOR = JavaLanguageLexer.LOR;
     /**
      * The {@code &&} (conditional AND) operator.
-     *
+     * <p>
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java:3636]: I recommend to move this change in separate PR to merge sooner and keep this PR small and focused on main issue

Comment (nrmancuso) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java]: There are a few more violations, so we can do this in a separate PR

**Comment by rnveach** [comment]
https://app.circleci.com/pipelines/github/checkstyle/checkstyle/15479/workflows/3f467140-9c31-40a3-9298-d3b602806eca/jobs/184239
CI Output:
````
Checking results ...
There are inspection problems. Review results at /home/circleci/project/target/inspection-results folder. Files:
/home/circleci/project/target/inspection-results/ParameterNameDiffersFromOverriddenParameter.xml
/home/circleci/project/target/inspection-results/JavadocBlankLines.xml
/home/circleci/project/target/inspection-results/JUnitMalformedDeclaration.xml

Exited with code exit status 1

CircleCI received exit code 1
````

Is this the expected output or will there be a follow up PR to show users violations?

**Comment by nrmancuso** [comment]
> https://app.circleci.com/pipelines/github/checkstyle/checkstyle/15479/workflows/3f467140-9c31-40a3-9298-d3b602806eca/jobs/184239 CI Output:
> 
> ```
> Checking results ...
> There are inspection problems. Review results at /home/circleci/project/target/inspection-results folder. Files:
> /home/circleci/project/target/inspection-results/ParameterNameDiffersFromOverriddenParameter.xml
> /home/circleci/project/target/inspection-results/JavadocBlankLines.xml
> /home/circleci/project/target/inspection-results/JUnitMalformedDeclaration.xml
> 
> Exited with code exit status 1
> 
> CircleCI received exit code 1
> ```
> 
> Is this the expected output or will there be a follow up PR to show users violations?

All reports are uploaded to the 'artifacts' tab: https://app.circleci.com/pipelines/github/checkstyle/checkstyle/15479/workflows/3f467140-9c31-40a3-9298-d3b602806eca/jobs/184239/artifacts

**Comment by romani** [comment]
>There are inspection problems. Review results at /home/circleci/project/target/inspection-results folder. Files:

Can we generate link to artifacts tab ?

**Comment by nrmancuso** [comment]
> > There are inspection problems. Review results at /home/circleci/project/target/inspection-results folder. Files:
> 
> Can we generate link to artifacts tab ?

We should be able to do this via built in environment variables, if not I will expense failure message at least.

**Comment by nrmancuso** [comment]
I was not able to generate a link to the artifacts tab, note the URL: https://app.circleci.com/pipelines/github/checkstyle/checkstyle/15489/workflows/28624edc-1df9-47db-a8b4-49e6c1675899/jobs/184452/artifacts

We can resolve the `CIRCLE_BUILD_URL` with curl , append to get close, but I cannot figure out where `15489` comes from (and this number changes every CI run). 

```bash
circleci@750f37235eb6:~$ printenv | grep 15489
circleci@750f37235eb6:~$ 

```

**Comment by romani** [comment]
ok, lets keep for now wording to look at "Atrifact" tab

**Comment by romani** [comment]
https://hub.docker.com/r/checkstyle/idea-docker/tags

File: .ci/idea_inspection.sh [hunk 4]
```diff
@@ -28,9 +28,6 @@ if [[ -z $IDEA_PATH ]]; then
     fi
 fi
 
-# Execute compilation of Checkstyle to generate all source files
-mvn -e --no-transfer-progress clean compile
-
```
Comment (nrmancuso) [inline .ci/idea_inspection.sh:33]: I have removed this from script since we ignore `target` directory during inspection anyway. This reduces execution time and makes execution in container locally easier.

File: .circleci/config.yml [hunk 5]
```diff
@@ -212,6 +212,25 @@ jobs:
...
+      - checkout
+      - run:
+          name: Run inspections
+          command: |
+            mkdir -p .idea
+            mv config/intellij-idea-inspections-misc.xml \
+              .idea/misc.xml
+            ls .idea
+            ./.ci/idea_inspection.sh
+          no_output_timeout: 25m
```
Comment (romani) [inline .circleci/config.yml:229]: Do we really expect that long execution ?
Can we just double of what we have now?

File: .circleci/config.yml [hunk 6]
```diff
@@ -212,6 +212,25 @@ jobs:
...
+      - image: checkstyle/idea-docker:jdk11-idea2022.2.2
+
+    steps:
+      - checkout
+      - run:
+          name: Run inspections
+          command: |
+            mkdir -p .idea
+            mv config/intellij-idea-inspections-misc.xml \
+              .idea/misc.xml
```
Comment (romani) [inline .circleci/config.yml:226]: Why not cp ? Can we make it single line command ?

File: .circleci/config.yml [hunk 7]
```diff
@@ -212,6 +212,25 @@ jobs:
...
+            ls .idea
```
Comment (romani) [inline .circleci/config.yml:227]: I think we can remove this debug output

File: .circleci/config.yml [hunk 8]
```diff
@@ -212,6 +212,25 @@ jobs:
...
+  run-inspections:
+    docker:
+      - image: checkstyle/idea-docker:jdk11-idea2022.2.2
+
+    steps:
+      - checkout
+      - run:
+          name: Run inspections
+          command: |
+            mkdir -p .idea
```
Comment (romani) [inline .circleci/config.yml:224]: Remove -p

File: config/intellij-idea-inspections-misc.xml [hunk 9]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (romani) [inline config/intellij-idea-inspections-misc.xml:1]: Why we have misc in name ?

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: config/intellij-idea-inspections-misc.xml (hunk 9 above)
Comment (nrmancuso) [reply config/intellij-idea-inspections-misc.xml:1]: This corresponds to name of file that inspections looks for, there should be some similarity here.

File: .ci/jsoref-spellchecker/whitelist.words [hunk 10]
```diff
@@ -979,6 +979,7 @@ npath
 npathcomplexity
 npe
 npm
+nrmancuso
```
Comment (romani) [inline .ci/jsoref-spellchecker/whitelist.words:982]: Can we removed, I think.

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: .circleci/config.yml (hunk 8 above)
Comment (nrmancuso) [reply .circleci/config.yml:224]: Why? We typically do this in all of our shell scripts, and CI will break if we decide to start checking in idea files in the future.

```
➜  checkstyle git:(issue-12087) ✗ grep -R "mkdir -p" | wc -l                 
25
➜  checkstyle git:(issue-12087) ✗ grep -R "mkdir" | grep -v "mkdir -p"
.github/workflows/pitest.yml:          mkdir staging && cp -r target/pit-reports/ staging
.ci/idea_inspection.bat:mkdir %RESULTS_DIR%
.ci/idea_inspection.bat:mkdir .idea\scopes
.ci/jsoref-spellchecker/whitelist.words:mkdir


```

File: config/intellij-idea-inspections-misc.xml (hunk 9 above)
Comment (romani) [reply config/intellij-idea-inspections-misc.xml:1]: Is it exact name that idea expects ? Or what do you mean by similarity?

File: .circleci/config.yml (hunk 8 above)
Comment (romani) [reply .circleci/config.yml:224]: It is required when we need to create whole path(few folders), not just a folder.

File: config/intellij-idea-inspections-misc.xml (hunk 9 above)
Comment (nrmancuso) [reply config/intellij-idea-inspections-misc.xml:1]:
```
➜  checkstyle git:(issue-12087) ✗ find config -name "intellij*" 
config/intellij-idea-inspections.xml
config/intellij-idea-inspections-misc.xml
config/intellij-idea-inspection-scope.xml
config/intellij-idea-inspections.properties
➜  checkstyle git:(issue-12087) ✗ ls .idea 
codeStyles    encodings.xml       jarRepositories.xml  scopes          vcs.xml
compiler.xml  inspectionProfiles  misc.xml             uiDesigner.xml  workspace.xml
```
All of the IDEA files that we check in to config directory have `intellij-idea-inspections*` prefix, and this file's original name is `misc.xml`.  So, it makes sense to prefix with common prefix, and keep original file name preserved for  others to understand relevance of file.

File: .circleci/config.yml (hunk 8 above)
Comment (nrmancuso) [reply .circleci/config.yml:224]: I did not use `-p` for parent usage, I used it to not fail if folder exists.

File: .circleci/config.yml (hunk 5 above)
Comment (romani) [reply .circleci/config.yml:229]: Ok, make sense, I remember you told me that execution is only few minutes by CLI :). Ok it is kind same as in TC, that is good.

Comment (nrmancuso) [reply .circleci/config.yml:229]: We can make execution faster by increasing resource class, if you want to. It ran in 4-5 minutes on large (default is medium for us).

File: .circleci/config.yml (hunk 7 above)
Comment (nrmancuso) [reply .circleci/config.yml]: removed

File: .ci/jsoref-spellchecker/whitelist.words (hunk 10 above)
Comment (nrmancuso) [reply .ci/jsoref-spellchecker/whitelist.words]: removed

File: .circleci/config.yml (hunk 8 above)
Comment (romani) [reply .circleci/config.yml:225]: But know for sure folder does not exist
https://github.com/checkstyle/contribution/commit/b95d0e4ffa008c77cb7a4a82796c2116bed5f1ec

File: config/intellij-idea-inspections-misc.xml (hunk 9 above)
Comment (romani) [reply config/intellij-idea-inspections-misc.xml:1]: https://github.com/checkstyle/checkstyle/tree/master/config

Ok, I am not about prefix, I am about suffix "- misc".
Can you name it "-project". Becase it is it is like project file for idea to know nature of it.

Comment (romani) [reply config/intellij-idea-inspections-misc.xml:1]: Or it is suffix that idea requires?
https://github.com/JetBrains/intellij-samples/blob/main/.idea/misc.xml

Comment (nrmancuso) [reply config/intellij-idea-inspections-misc.xml:1]: common prefix for all idea config: intellij-idea-inspections
actual file name that idea requires: misc.xml

File: .ci/idea_inspection.sh (hunk 4 above)
Comment (nrmancuso) [reply .ci/idea_inspection.sh]: I think removal of this is why inspections are failing, testing now

File: .circleci/config.yml (hunk 8 above)
Comment (nrmancuso) [reply .circleci/config.yml]: removed

**Review by romani — APPROVED** [review]
Ok to merge

**Comment by nrmancuso** [comment]
> > issue is resolved, project build is required before inspections execution. I added a comment to make this obvious.
> 
> Can it be a reason of problem in TC ?

I was thinking the same, it is likely that this is the issue. Not sure if it is worth investigating, I for one would be happy to never deal with TC interface again.

**Comment by romani** [comment]
Extra CI is always good.

**Comment by rnveach** [comment]
> Extra CI is always good.

Red CI is not good. TC has been red for a while now (feels like weeks).

**Comment by nrmancuso** [comment]
TC has three strikes for me:
1. Poor UI and difficult to use
2. In project config is buggy/weird
3. We must reach out to support for most issues

This PR puts control back in our hands of updates/ support and allows all configuration to be in the project.

This is enough for me to vote it out of CI completely.

**Comment by nrmancuso** [comment]
Yes, we will be failing until violations are fixed/ suppressed, we can do this in separate PRs

**Comment by rnveach** [comment]
> `<problem_class id="JUnitMalformedDeclaration" severity="WARNING"`

So all warnings are really errors?

**Comment by nrmancuso** [comment]
We fail on any problems, this could be warnings or errors.

File: config/intellij-idea-inspections.properties [hunk 11]
```diff
@@ -2,7 +2,10 @@
 # Content should be the same as in TeamCity "Include / exclude patterns:"
 idea.exclude.patterns=.idea/**;\
               .ci/*.config;\
-              .ci/pitest-survival-check-xml.groovy\
+              .ci/checker-framework.groovy;\
+              .ci/codenarc.groovy;\
+              .ci/error-prone-check.groovy;\
+              .ci/pitest-survival-check-xml.groovy;\
```
Comment (rnveach) [inline config/intellij-idea-inspections.properties:8]: Curious, what happens if we run this on groovy scripts? Is this a future possibility / issue?

**Review by rnveach — APPROVED** [review]
@romani ping https://github.com/checkstyle/checkstyle/pull/12188#issuecomment-1270831064

File: config/intellij-idea-inspections.properties (hunk 11 above)
Comment (nrmancuso) [reply config/intellij-idea-inspections.properties:8]: Yes, we get a few weird violations due to not having groovy configured correctly (I think it is a language level issue)

**Comment by romani** [comment]
We can detach TC from CI status for now, but keep it in case we have more work for it.

**Comment by romani** [comment]
New job should have green state to be merged.

**Comment by rnveach** [comment]
@romani 
> We can detach TC from CI status for now, but keep it in case we have more work for it.

How do we do this?

**Comment by nrmancuso** [comment]
I will make one last effort to get TC working, let’s see how this run works out.

**Comment by nrmancuso** [comment]
TC is still failing even with `mvn clean compile` step first as we do in inspections script. https://teamcity.jetbrains.com/buildConfiguration/Checkstyle_IdeaInspectionsPullRequest/3967318?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true&expandBuildProblemsSection=true

**Comment by romani** [comment]
mistake of this PR is  .... why we choose not same version as we use in TC ?

**Comment by romani** [comment]
I pushed commit with suppression and fixes from https://github.com/checkstyle/checkstyle/pull/12274
